src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3068,6 +3068,118 @@ public override MemberInfo[] GetMember(string name, MemberTypes type, BindingFla
 
             return compressMembers;
         }
+
+        public override MemberInfo? GetMemberWithSameMetadataDefinitionAs(MemberInfo member)
+        {
+            if (member is null) throw new ArgumentNullException(nameof(member));
+
+            return member.MemberType switch
+            {
+                MemberTypes.Method => GetMethodWithSameMetadataDefinitionAs(member),
+                MemberTypes.Constructor => GetConstructorWithSameMetadataDefinitionAs(member),
+                MemberTypes.Property => GetPropertyWithSameMetadataDefinitionAs(member),
+                MemberTypes.Field => GetFieldWithSameMetadataDefinitionAs(member),
+                MemberTypes.Event => GetEventWithSameMetadataDefinitionAs(member),
+                MemberTypes.NestedType => GetNestedTypeWithSameMetadataDefinitionAs(member),
+                _ => null
```

Comment:
It looks odd to me that we allow people to pass in unrelated methods; return null when they do that; and then force every caller to do `!` on the result hoping that they got it right.

Should this throw exception instead of returning null?

Reply:
My thinking was that it behaves the same as all the other `GetXXX` methods on `Type`. You can call `typeof(string).GetMethod("Foo")` and it doesn't throw, but instead returns `null`. This behaves the same if you pass in an incompatible member.

Reply:
It is common to lookup method by name and expect it to not be there. Also, `GetMethod` is still going to throw when you pass it invalid input and the match is ambiguous.

I do not  think there is any reasonable scenario where somebody would want to pass MemberInfo from unrelated type into this API and expect it to work. This API should treat MemberInfo from unrelated type as invalid input and throw for it. Once it throws for invalid input, it will never return null.

Reply:
Are we concerned at all that validating the input is just going to slow down the happy path?

Reply:
I think it is fine to throw the exception after the fact.